Integrate Sherrington-Kirkpatrick (SK) model

bgupt opened the pull request:
1. Adding SK model objective function.
2. Writing tests for SK model.

rsln-s commented on qokit/qaoa_circuit_sk.py:
```diff
@@ -0,0 +1,139 @@
+from typing import Sequence
+
+
+def append_zz_term(qc, q1, q2, gamma):
```
Appears to duplicate the function in qokit/qaoa_circuit_maxcut.py; idea: create qaoa_circuit_utils.py and import from there. Same for others below

bgupt replied:
Done!

rsln-s commented on qokit/qaoa_circuit_sk.py:
```diff
@@ -0,0 +1,139 @@
+    return qc
+
+
+def get_parameterized_qaoa_circuit(
```
Now that there's more of them, it may be a good time to unify the constructors

rsln-s commented on examples/benchmark.ipynb:
Checkout from main to remove the diff (same for other notebooks)

bgupt replied:
Done!

rsln-s commented on qokit/yahoo.py:
Brajesh says this change can be undone

rsln-s commented on qokit/sk.py:
```diff
@@ -0,0 +1,64 @@
+    """Compute the value of a cut.
+    Args:
+        x (numpy.ndarray): binary string as numpy array.
+        w (numpy.ndarray): adjacency matrix returned by get_adjacency_matrix
```
Not an adjacency matrix, not `w`. Should be something like "matrix specifying the couplings in the SK model"

bgupt replied:
Done!

rsln-s commented on qokit/sk.py:
```diff
@@ -0,0 +1,64 @@
+    return terms
+
+
+def get_adjacency_matrix(G: nx.Graph) -> np.ndarray:
```
Not needed

rsln-s commented on qokit/sk.py:
```diff
@@ -0,0 +1,64 @@
+        terms to be used in the simulation
+    """
+    N = J.shape[0]
+    G = nx.complete_graph(N)
```
Graph is not needed

rsln-s commented on tests/test_qaoa_objective_sk.py:
```diff
@@ -0,0 +1,213 @@
+        get_qaoa_sk_objective(N, p, J=J, parameterization="gamma beta", simulator=simulator, objective="expectation")(gamma, beta)
+        for simulator in simulators_to_run_names_no_qiskit
+    ]
+    print(qaoa_objectives)
```
Please remove debugging prints!

rsln-s commented on tests/test_qaoa_objective_sk.py:
```diff
@@ -0,0 +1,213 @@
+
+    obj = partial(sk_obj, J=J)
+    precomputed_energies = precompute_energies(obj, N)
+    optimal_cut = np.max(precomputed_energies)
```
Why "`optimal_cut`"? It's not really a cut, is it?

bgupt replied:
Done!

rsln-s commented on tests/test_qaoa_objective_sk.py:
```diff
@@ -0,0 +1,213 @@
+
+def test_sk_qaoa_convergence_with_p():
+    N = 8
+    J = np.random.randn(N, N)
```
From [numpy documentation](https://numpy.org/doc/2.2/reference/random/generated/numpy.random.randn.html): "New code should use the [standard_normal](https://numpy.org/doc/2.2/reference/random/generated/numpy.random.Generator.standard_normal.html#numpy.random.Generator.standard_normal) method of a [Generator](https://numpy.org/doc/2.2/reference/random/generator.html#numpy.random.Generator) instance instead; please see the [Quick start](https://numpy.org/doc/2.2/reference/random/index.html#random-quick-start)."

rsln-s commented on tests/test_qaoa_objective_sk.py:
```diff
@@ -0,0 +1,213 @@
+            ]
+            print("p: ", p, "Approximation Ratio:", qaoa_objectives / optimal_cut)
+            current_ar = qaoa_objectives / optimal_cut
+            assert list(current_ar) > list(last_ar)
```
Why cast to `list` here?

rsln-s commented on tests/test_qaoa_objective_sk.py:
```diff
@@ -0,0 +1,213 @@
+    J = (J + J.T) / 2
+    np.fill_diagonal(J, 0)
+
+    terms = [(2 * J[spin_pair[0], spin_pair[1]] / np.sqrt(N), spin_pair) for spin_pair in combinations(range(N), r=2)]
```
Is there a `get_sk_terms` function?

bgupt replied:
Fixed!

rsln-s commented on tests/test_qaoa_objective_sk.py:
```diff
@@ -0,0 +1,213 @@
+    J = np.random.randn(N, N)
+    J = (J + J.T) / 2
+    np.fill_diagonal(J, 0)
+    for max_p in [11, 4]:
```
The purpose of this for loop is unclear to me; why vary max_p if running the inside of the loop with `max_p=4` corresponds to a subset of the run with `max_p=11`?

bgupt replied:
Fixed!

rsln-s commented on tests/test_qaoa_objective_sk.py:
```diff
@@ -0,0 +1,213 @@
+qiskit_backend = AerSimulator(method="statevector")
+SIMULATORS = get_available_simulators("x") + get_available_simulators("xyring") + get_available_simulators("xycomplete")
+simulators_to_run_names = get_available_simulator_names("x") + ["qiskit"]
+simulators_to_run_names_no_qiskit = get_available_simulator_names("x")
```
In all of the below tests qiskit is excluded (only `simulators_to_run_names_no_qiskit` is used). Why?

rsln-s commented on setup.py:
```diff
@@ -27,7 +27,7 @@
         extensions.append(Extension("simulator", sources=sources, include_dirs=[os.path.join(path, "")], extra_compile_args=["/d2FH4-"]))
     elif sys.platform.startswith("darwin"):
-        extensions.append(Extension("simulator", sources=sources, include_dirs=[os.path.join(path, "")], extra_compile_args=["-Xpreprocessor"]))
+        extensions.append(Extension("simulator", language="c", sources=sources, include_dirs=[os.path.join(path, "")], extra_compile_args=["-Xpreprocessor"]))
```
Could you clarify why this change is needed?

bgupt replied:
The option language is not required as one would ned to append compiler flag for clang anyway at the end when compiling with 'c'.

rsln-s commented on tests/test_qaoa_objective_sk.py:
```diff
@@ -0,0 +1,187 @@
+    precomputed_energies = precompute_energies(obj, N)
+    max_energy = np.max(precomputed_energies)
+
+    last_ar = [0.0, 0.0]
```
If there's a reason to manually iterate over the simulators, should this be `len(simulators_to_run_names) * [0.0]`?

bgupt replied:
Yes, indeed! Fixed.

bgupt replied:
Parametrized the test with "simulator"!

rsln-s commented on tests/test_qaoa_objective_sk.py:
```diff
@@ -0,0 +1,187 @@
+    assert np.allclose(qaoa_objectives, expected_energy)
+
+
+def test_sk_qaoa_convergence_with_p():
```
Why not use `@pytest.mark.parametrize("simulator", simulators_to_run_names)` for this test instead of manually iterating over the simulators?

bgupt replied:
Done!

rsln-s commented on tests/test_qaoa_objective_sk.py:
```diff
@@ -0,0 +1,187 @@
+    assert np.isclose(N * var, 1, atol=1e-2)
+
+
+def test_energy_pre_optimized():
```
Why not use `@pytest.mark.parametrize("simulator", simulators_to_run_names)` for this test instead of manually iterating over the simulators?

bgupt replied:
Done!

rsln-s commented on tests/test_qaoa_objective_sk.py:
```diff
@@ -0,0 +1,187 @@
+
+rng = np.random.default_rng(seed=42)
+
+qiskit_backend = AerSimulator(method="statevector")
```
`qiskit_backend`, not used anywhere in this file

bgupt replied:
Removed!

rsln-s commented on tests/test_qaoa_objective_sk.py:
```diff
@@ -0,0 +1,187 @@
+rng = np.random.default_rng(seed=42)
+
+qiskit_backend = AerSimulator(method="statevector")
+SIMULATORS = get_available_simulators("x") + get_available_simulators("xyring") + get_available_simulators("xycomplete")
```
Something is off; nothing in this test file should have `xy` mixer.

bgupt replied:
Removed `xy` mixers!

rsln-s commented on tests/test_qaoa_objective_sk.py:
```diff
@@ -0,0 +1,187 @@
+simulators_to_run_names_no_qiskit = get_available_simulator_names("x")
+
+
+def test_sk_obj(N=5):
```
Please make how N is set consistent. In some tests it is an argument, and in some tests it is set on the first line of the test

bgupt replied:
Done!

rsln-s commented on qokit/qaoa_circuit_utils.py:
```diff
@@ -0,0 +1,291 @@
+
+
+def append_cost_operator_circuit(qc, terms, gamma):
+    for term, (i, j) in terms:
```
Would be more clear if we used variable name `coeff` instead of `term`,  i.e. `for coeff, (i, j) in terms:`

bgupt replied:
Done!

rsln-s reviewed (changes requested):
Looks great! The only outstanding comment I have is regarding the "language" keyword option in setup.py; once that is resolved, should be good to merge

bgupt commented:
Thanks a lot for the email. I previously misread your comment about the language option. For some reason, I thought you were asking why I removed this option which is what I meant to do. I had added this to test out C compiling  on MacOS, but I suppose I did not remove it before the final commit. So, we are on the same page about the fact that the ‘language’ option should not be there. We would need to create a separate PR to tackle the C compiler issue on MacOS. I will revert these changes right away.

Sorry for the confusion!

-Brajesh

From: Ruslan Shaydulin ***@***.***>
Date: Sunday, March 9, 2025 at 1:24 PM
To: jpmorganchase/QOKit ***@***.***>
Cc: Gupt, Brajesh ***@***.***>, Author ***@***.***>
Subject: [EXTERNAL]Re: [jpmorganchase/QOKit] Integrate Sherrington-Kirkpatrick (SK) model (PR #82)

@rsln-s requested changes on this pull request.

Looks great! The only outstanding comment I have is regarding the "language" keyword option in setup.py; once that is resolved, should be good to merge

________________________________

In setup.py<https://github.com/jpmorganchase/QOKit/pull/82#discussion_r1986357433>:

> @@ -27,7 +27,7 @@

     if sys.platform in ["win32"]:

         extensions.append(Extension("simulator", sources=sources, include_dirs=[os.path.join(path, "")], extra_compile_args=["/d2FH4-"]))

     elif sys.platform.startswith("darwin"):

-        extensions.append(Extension("simulator", sources=sources, include_dirs=[os.path.join(path, "")], extra_compile_args=["-Xpreprocessor"]))

+        extensions.append(Extension("simulator", language="c", sources=sources, include_dirs=[os.path.join(path, "")], extra_compile_args=["-Xpreprocessor"]))

I think I don't follow... If the option "language" is not required, why do we add it in this PR? Should this change be removed?

—
Reply to this email directly, view it on GitHub<https://github.com/jpmorganchase/QOKit/pull/82#pullrequestreview-2669461080>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABXVWBPOHOZDIULMY2YNLFL2TSBPBAVCNFSM6AAAAABWDULJP6VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMNRZGQ3DCMBYGA>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

This message is confidential and subject to terms at: https://www.jpmorgan.com/emaildisclaimer including on confidential, privileged or legal entity information, malicious content and monitoring of electronic messages. If you are not the intended recipient, please delete this message and notify the sender immediately. Any unauthorized use is strictly prohibited.

rsln-s reviewed (approved):
LGTM!
